File: backend/invites.py
"""Secure one-time staff invitations.

A Super Admin creates the staff account without a password. A single-use setup token is
generated, stored only as a hash, expires after INVITE_TTL_HOURS, and is invalidated when it is
used or when a new invite is issued. No password is ever chosen or seen by the Super Admin.
"""
import hashlib
import secrets
import uuid
from datetime import datetime, timedelta, timezone

from db import db

INVITE_TTL_HOURS = 72


def _hash(token: str) -> str:
    return hashlib.sha256(token.encode("utf-8")).hexdigest()


async def issue_invite(user_id: str, email: str, invited_by: str) -> dict:
    """Invalidate any previous invite for this user and issue a fresh single-use token."""
    now = datetime.now(timezone.utc)
    await db.staff_invites.update_many(
        {"user_id": user_id, "used_at": None, "revoked_at": None},
        {"$set": {"revoked_at": now.isoformat()}})
    token = secrets.token_urlsafe(32)
    expires = now + timedelta(hours=INVITE_TTL_HOURS)
    await db.staff_invites.insert_one({
        "id": str(uuid.uuid4()), "user_id": user_id, "email": email,
        "token_hash": _hash(token), "invited_by": invited_by,
        "expires_at": expires.isoformat(), "used_at": None, "revoked_at": None,
        "created_at": now.isoformat(),
    })
    return {"token": token, "expires_at": expires.isoformat()}


async def find_valid_invite(token: str) -> dict | None:
    invite = await db.staff_invites.find_one({"token_hash": _hash(token)})
    if not invite or invite.get("used_at") or invite.get("revoked_at"):
        return None
    if invite["expires_at"] < datetime.now(timezone.utc).isoformat():
        return None
    return invite


async def consume_invite(invite_id: str) -> None:
    await db.staff_invites.update_one(
        {"id": invite_id}, {"$set": {"used_at": datetime.now(timezone.utc).isoformat()}})

File: backend-node/src/services/invites.ts
/**
 * Secure one-time staff invitations.
 *
 * A Super Admin creates the staff account without a password. A single-use setup token is
 * generated, stored only as a hash, expires after INVITE_TTL_HOURS, and is invalidated when it
 * is used or when a new invite is issued. No password is ever chosen or seen by the Super Admin.
 */
import { createHash, randomBytes, randomUUID } from "crypto";

import { col, Doc } from "../db/mongo";

export const INVITE_TTL_HOURS = 72;

function hashToken(token: string): string {
  return createHash("sha256").update(token, "utf8").digest("hex");
}

/** Invalidate any previous invite for this user and issue a fresh single-use token. */
export async function issueInvite(
  userId: string,
  email: string,
  invitedBy: string,
): Promise<{ id: string; token: string; expires_at: string }> {
  const now = new Date();
  await col("staff_invites").updateMany(
    { user_id: userId, used_at: null, revoked_at: null },
    { $set: { revoked_at: now.toISOString() } },
  );
  const token = randomBytes(32).toString("base64url");
  const expires = new Date(now.getTime() + INVITE_TTL_HOURS * 3600 * 1000);
  const id = randomUUID();
  await col("staff_invites").insertOne({
    id,
    user_id: userId,
    email,
    token_hash: hashToken(token),
    invited_by: invitedBy,
    expires_at: expires.toISOString(),
    used_at: null,
    revoked_at: null,
    created_at: now.toISOString(),
  });
  return { id, token, expires_at: expires.toISOString() };
}

export async function findValidInvite(token: string): Promise<Doc | null> {
  const invite = (await col("staff_invites").findOne({ token_hash: hashToken(token) })) as Doc | null;
  if (!invite || invite.used_at || invite.revoked_at) return null;
  if (invite.expires_at < new Date().toISOString()) return null;
  return invite;
}

export async function consumeInvite(inviteId: string): Promise<void> {
  await col("staff_invites").updateOne(
    { id: inviteId },
    { $set: { used_at: new Date().toISOString() } },
  );
}
